controlplane/kubeadm/internal/etcd_client_generator.go

```diff
@@ -34,22 +34,41 @@ import (
 type etcdClientGenerator struct {
 	restConfig *rest.Config
 	tlsConfig  *tls.Config
+	clientFactory
 }
 
-func (c *etcdClientGenerator) forNodes(ctx context.Context, nodes []corev1.Node) (*etcd.Client, error) {
-	endpoints := make([]string, len(nodes))
-	for i, node := range nodes {
-		endpoints[i] = staticPodName("etcd", node.Name)
+type clientFactory func(ctx context.Context, endpoints []string) (*etcd.Client, error)
+
+// getClientFactory returns a clientFactory. sets a default factory if not present
+func (c *etcdClientGenerator) getClientFactory() clientFactory {
+	if c.clientFactory == nil {
+		c.clientFactory = func(ctx context.Context, endpoints []string) (*etcd.Client, error) {
+			p := proxy.Proxy{
+				Kind:       "pods",
+				Namespace:  metav1.NamespaceSystem,
+				KubeConfig: c.restConfig,
+				TLSConfig:  c.tlsConfig,
+				Port:       2379,
+			}
+			return etcd.NewClient(ctx, endpoints, p, c.tlsConfig)
+		}
 	}
+	return c.clientFactory
+}
 
-	p := proxy.Proxy{
-		Kind:       "pods",
-		Namespace:  metav1.NamespaceSystem,
-		KubeConfig: c.restConfig,
-		TLSConfig:  c.tlsConfig,
-		Port:       2379,
+// forNodes takes a list of nodes and returns a client for the first one that connects
+func (c *etcdClientGenerator) forNodes(ctx context.Context, nodes []corev1.Node) (*etcd.Client, error) {
```

Comment:
Sidenote, this method name is super confusing and I had to read the comment above to understand what it does. I originally assumed that `forNodes` would return a client connecting (pooling) to all of the pods.

How about renaming it to something like `forFirstAvailable`? Or something similar, not sure which one is better at this points 😄 

cc @sedefsavas @randomvariable

Reply:
+1 for renaming

Reply:
Ditto